optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/change/DefaultProblemChangeDirector.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.solver.change;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import org.optaplanner.core.api.score.director.ScoreDirector;
+import org.optaplanner.core.api.solver.change.ProblemChangeDirector;
+
+public final class DefaultProblemChangeDirector<Solution_> implements ProblemChangeDirector {
+
+    private final ScoreDirector<Solution_> scoreDirector;
+
+    public DefaultProblemChangeDirector(ScoreDirector<Solution_> scoreDirector) {
+        this.scoreDirector = scoreDirector;
+    }
+
+    @Override
+    public <Entity> void addEntity(Entity entity, Consumer<Entity> entityConsumer) {
+        Objects.requireNonNull(entity, "Entity (" + entity + ") cannot be null.");
+        Objects.requireNonNull(entityConsumer, "Entity consumer (" + entityConsumer + ") cannot be null.");
+        scoreDirector.beforeEntityAdded(entity);
+        entityConsumer.accept(entity);
+        scoreDirector.afterEntityAdded(entity);
+    }
+
+    @Override
+    public <Entity> void removeEntity(Entity entity, Consumer<Entity> entityConsumer) {
+        Objects.requireNonNull(entity, "Entity (" + entity + ") cannot be null.");
+        Objects.requireNonNull(entityConsumer, "Entity consumer (" + entityConsumer + ") cannot be null.");
+        Entity workingEntity = lookUpWorkingObject(entity);
+        scoreDirector.beforeEntityRemoved(workingEntity);
+        entityConsumer.accept(workingEntity);
+        scoreDirector.afterEntityRemoved(workingEntity);
+    }
+
+    @Override
+    public <Entity> void changeVariable(Entity entity, String variableName, Consumer<Entity> entityConsumer) {
+        Objects.requireNonNull(entity, "Entity (" + entity + ") cannot be null.");
+        Objects.requireNonNull(variableName, "Planning variable name (" + variableName + ") cannot be null.");
+        Objects.requireNonNull(entityConsumer, "Entity consumer (" + entityConsumer + ") cannot be null.");
```

Comment:
soft: these string concats for every one of these changevariable calls (typicaly done in a loop) can add up